### PR TITLE
Create @automattic/sentry package

### PR DESCRIPTION
**Pull request by p-jackson**
#### Proposed Changes

Moving the code at `client/lib/sentry` to `@automattic/sentry` so that other packages can send messages and exceptions to Sentry.

This idea spawned from #66061 where [I wanted to catch an asynchronous exception and make sure it got funneled to Sentry](https://github.com/Automattic/wp-calypso/pull/66061#discussion_r938463912). The functions in `client/lib/sentry` allow this, however the code where the exception is handled is in `@automattic/components` and so it can't import the Sentry library.

This PR creates a new `@automattic/sentry` package and copies the client lib code into it.

There's probably a way to decouple the `initSentry()` code from the `captureException()` code so that users of this package only have to bundle the bare minimum of code. But I think this is a good simple first approach.

#### Testing Instructions

*

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #66061

### CODE REVIEW

**Comment by noahtallen** [comment]
This is really interesting, thanks for working on it!

My concern is that Sentry is by definition tied to the Calypso application. The config we have is specific to running on `wordpress.com`, etc. This config shouldn't be used in wp-admin, in ETK, for example.

But if it's exposed as a package, wouldn't that allow for a scenario where package A uses the sentry lib, ETK uses package A, and then we get two instances of Sentry running in wp-admin? (ETK already bootstraps its own instance.)

Plus, if we're using the components package outside of Calypso across Automattic, we probably don't want calypso's sentry version to be loaded there. (And we definitely don't want that for 3rd party consumers.)

I'm not sure if there's an easy way to solve that with this approach. 🤔

**Comment by p-jackson** [comment]
Thanks @noahtallen those are good points. The `initSentry` is binding event listeners to `window`, so it really is more of an app-level concern. The reason I think it's worth trying is that the alternative is a lot of prop drilling: if a component in `@automattic/components` might raise an error then it'll have to accept an `onError` prop, and the Calypso code will have to pass the `captureException` function all the way down through various `onError` props. I don't think many people could be bothered doing that, so I want it to be as easy as possible to capture the exception.

I've given it another try
- `@automattic/sentry` only has method shims
- most of the code (especially Calypso config) is still in `lib/sentry`
- coupling between the package and `lib/sentry` is as loose as possible

So if the `captureException` function is called outside a Calypso environment it's a noop.

If we only had to support React then another idea is to use `context` to provide the `dispatchSentryMethodCall` method. That'd address the prop drilling case I mentioned above. But the way I've done it is useful for non-React situations.

**Comment by noahtallen** [comment]
Yeah, that's a good point. We did use filters for a similar use case (capturing Gutenberg errors within React error boundaries), but that was really the only viable solution for integration with Gutenberg.

A benefit of this approach is that we can define `window.a8cSentryCallQueue` in our other environments (like ETK in wp-admin), so this would work with whichever Sentry implementation is available

File: packages/sentry/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+function dispatchSentryMethodCall< Method extends SupportedMethods >(
+	method: Method,
+	args: Parameters< typeof SentryApi[ Method ] >
+) {
+	if ( window.a8cDispatchSentryMethodCall ) {
+		window.a8cDispatchSentryMethodCall( method, args );
+	} else {
+		window.a8cSentryCallQueue = window.a8cSentryCallQueue || [];
+		window.a8cSentryCallQueue.push( { f: method, a: args } );
```
Comment (noahtallen) [inline packages/sentry/src/index.ts:36]: I think this will grow the call queue indefinitely if Sentry is never loaded in an environment.

Comment (noahtallen) [reply packages/sentry/src/index.ts:36]: The reason this wasn't an issue in the other file is that we clear the call queue there if Sentry doesn't get loaded (which works when that file is also in charge of loading Sentry).

Comment (p-jackson) [reply packages/sentry/src/index.ts:36]: Ah yep, you're right. If `status` was `'error'` or `'disabled'` we'd stop pushing items onto the array. I should be able to fix that with yet another global 😆 

Although maybe I should have a go using Gutenberg filters instead like you mentioned above.

Comment (noahtallen) [reply packages/sentry/src/index.ts:36]: That said, if there is no sentry environment at all -- like if `@automattic/components` was used in a different application -- those statuses wouldn't get populated

**Comment by p-jackson** [comment]
I'm not quite sure about the cleanest way to do this. I'll just leave it for now. Next time a need comes up it might be more obvious what to do.

**Comment by noahtallen** [comment]
Ok! It's definitely an interesting problem. Maybe at some point we (@Automattic/team-calypso-platform) could think about how to have a sharable Sentry library which can connect to the running sentry instance on any platform :)

**Pull request closed without merging**